extensions/cloudfiles-extensions/src/main/java/io/druid/storage/cloudfiles/CloudFilesDataSegmentPuller.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Druid - a distributed column store.
+ * Copyright 2012 - 2015 Metamarkets Group Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.druid.storage.cloudfiles;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Map;
+
+import org.jclouds.rackspace.cloudfiles.v1.CloudFilesApi;
+
+import com.google.inject.Inject;
+import com.metamx.common.CompressionUtils;
+import com.metamx.common.FileUtils;
+import com.metamx.common.ISE;
+import com.metamx.common.MapUtils;
+import com.metamx.emitter.EmittingLogger;
+
+import io.druid.segment.loading.DataSegmentPuller;
+import io.druid.segment.loading.SegmentLoadingException;
+import io.druid.timeline.DataSegment;
+
+/**
+ * A data segment puller that also handles URI data pulls.
+ */
+public class CloudFilesDataSegmentPuller implements DataSegmentPuller {
+
+	public static final String scheme = CloudFilesStorageDruidModule.SCHEME;
+
+	private static final EmittingLogger log = new EmittingLogger(CloudFilesDataSegmentPuller.class);
+
+	private final CloudFilesApi cloudFilesApi;
+
+	@Inject
+	public CloudFilesDataSegmentPuller(final CloudFilesApi cloudFilesApi) {
+		this.cloudFilesApi = cloudFilesApi;
+	}
+
+	@Override
+	public void getSegmentFiles(final DataSegment segment, final File outDir) throws SegmentLoadingException {
+		Map<String, Object> loadSpec = segment.getLoadSpec();
+		String region = MapUtils.getString(loadSpec, "region");
+		String container = MapUtils.getString(loadSpec, "container");
+		String path = MapUtils.getString(loadSpec, "path");
+
+		log.info("Pulling index at path[%s] to outDir[%s]", path, outDir);
+		prepareOutDir(outDir);
+		getSegmentFiles(region, container, path, outDir);
+	}
+
+	public FileUtils.FileCopyResult getSegmentFiles(String region, String container, String path, File outDir)
+			throws SegmentLoadingException {
+		CloudFilesObjectApiProxy objectApi = new CloudFilesObjectApiProxy(cloudFilesApi, region, container);
+		CloudFilesObject cloudFilesObject = objectApi.get(path);
+
+		try {
+			final InputStream inputStream = cloudFilesObject.getPayload().openStream();
```

Comment:
Thanks for the tip! I changed it by adding a `CloudFilesByteSource` similarly to `AzureByteSource`. I'm gonna try if it's working.